hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfs/jni_helper.c

```diff
@@ -534,6 +545,21 @@ JNIEnv* getJNIEnv(void)
 
     state->env = getGlobalJNIEnv();
     mutexUnlock(&jvmMutex);
+
+    if (!jclassesInitialized) {
+      mutexLock(&jclassInitMutex);
+      if (!jclassesInitialized) {
```

Comment:
this double-checked locking idiom isn't safe unless you insert a memory barrier before the store to jclassesInitialized. (it has to be a store with "release" semantics)

Otherwise, the compiler (and/or CPU) is free to decide to reorder the store to 'jclassesInitialized = 1' up earlier within the critical section.

Reply:
that said, this isn't a hot path so I don't think it's worth trying to be fancy, and I think we can just acquire the mutex unconditionally

Reply:
Fixed. Removed the first `if` statement, so `initCachedClasses` always acquires the lock.